File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilNagumoCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.brasil;

import br.com.lett.crawlernode.core.fetcher.ProxyCollection;
import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.fetcher.models.Response;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.crawlers.extractionutils.core.SiteMercadoCrawler;
import br.com.lett.crawlernode.util.CommonMethods;
import org.apache.http.HttpHeaders;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class BrasilNagumoCrawler extends SiteMercadoCrawler {

   private final String API_URL = "https://www.nagumo.com.br/api/b2c/";

   private Map<String, Integer> lojaInfo = getLojaInfo();
   private Double latitude = session.getOptions().optDouble("latitude");
   private Double longitude = session.getOptions().optDouble("longitude");

   public BrasilNagumoCrawler(Session session) {
      super(session);
   }

   @Override
   protected Map<String, Integer> getLojaInfo() {
      return super.getLojaInfo();
   }

   @Override
   protected Response fetchResponse() {
      return crawlProductInformatioFromApi(session.getOriginalURL());
   }

   @Override
   protected Response crawlProductInformatioFromApi(String productUrl) {
      String productName = CommonMethods.getLast(productUrl.split("/")).split("\\?")[0];
      String url = API_URL + "product/" + productName + "?store_id=" + lojaInfo.get("IdLoja");

      Map<String, String> headers = new HashMap<>();
      headers.put("hosturl", "www.nagumo.com.br");
      headers.put(HttpHeaders.ACCEPT, "application/json, text/plain, */*");
      headers.put("sm-token", "{\"Location\":{\"Latitude\":" + latitude + ",\"Longitude\":" + longitude + "},\"IdLoja\":" + lojaInfo.get("IdLoja") + ",\"IdRede\":" + lojaInfo.get("IdRede") + "}");

      Request requestApi = Request.RequestBuilder.create()
         .setUrl(url)
         .setCookies(cookies)
         .setHeaders(headers)
         .setProxyservice(Arrays.asList(ProxyCollection.BUY, ProxyCollection.LUMINATI_RESIDENTIAL_BR, ProxyCollection.NETNUT_RESIDENTIAL_ES_HAPROXY))
         .build();

      return this.dataFetcher.get(session, requestApi);
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilNagumoCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.brasil

import br.com.lett.crawlernode.core.fetcher.methods.ApacheDataFetcher
import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.crawlers.extractionutils.core.BrasilSitemercadoCrawler
import br.com.lett.crawlernode.util.CrawlerUtils
import br.com.lett.crawlernode.util.JSONUtils
import org.apache.http.HttpHeaders
import org.json.JSONObject
import java.net.URI

class BrasilNagumoCrawler(session: Session) : BrasilSitemercadoCrawler(session) {

   companion object {
      private const val HOME_PAGE = "https://www.nagumo.com.br/guarulhos-lj42-guarulhos-aruja-jardim-cumbica-caminho-do-campo-do-rincao"
      private const val API_URL = "https://b2c-api-premiumlabel-production.azurewebsites.net/api/"

      private const val IDLOJA = 4951
      private const val IDREDE = 884
   }

   override fun getHomePage(): String {
      return HOME_PAGE
   }

   override fun getLojaInfo(): Map<String, Int> {
      val lojaInfo: MutableMap<String, Int> = HashMap()
      lojaInfo["IdLoja"] = IDLOJA
      lojaInfo["IdRede"] = IDREDE
      return lojaInfo
   }

   override fun getApiUrl(): String {
      return API_URL
   }

   override fun getLoadApiUrl(): String {
      return API_URL
   }

   override fun crawlProductInformatioFromApi(productUrl: String): JSONObject? {
      val lojaUrl = URI.create(session.originalURL).path.split("/").first(String::isNotEmpty)
      val loadUrl = "${API_URL}v1/b2c/page/store/$lojaUrl"
      val url = "${API_URL}b2c/product/${productUrl.split("/").last().split("\\?").first()}?store_id=${lojaInfo["IdLoja"]}"
      val headers: MutableMap<String, String?> = mutableMapOf(
         HttpHeaders.REFERER to productUrl,
         HttpHeaders.ACCEPT to "application/json, text/plain, */*",
         HttpHeaders.CONTENT_TYPE to "application/json",
         HttpHeaders.USER_AGENT to "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36"
      )
      val request = Request.RequestBuilder.create().setUrl(loadUrl).setCookies(cookies).setHeaders(headers).setPayload(loadPayload).build()
      val responseHeaders = ApacheDataFetcher()[session, request].headers

      val jsonObject = if (responseHeaders != null) JSONUtils.stringToJson(responseHeaders["sm-token"]) else JSONObject()
      jsonObject.put("IdLoja", lojaInfo["IdLoja"])
         .put("IdRede", lojaInfo["IdRede"])
      headers["sm-token"] = jsonObject.toString()
      headers["sm-mmc"] = responseHeaders["sm-mmc"]
      val requestApi = Request.RequestBuilder.create().setUrl(url).setCookies(cookies).setHeaders(headers).build()
      return CrawlerUtils.stringToJson(dataFetcher[session, requestApi].body)
   }

}
